src/components/views/settings/AddRemoveThreepids.tsx

```diff
@@ -0,0 +1,534 @@
+/*
+Copyright 2024 The Matrix.org Foundation C.I.C.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import React, { useCallback, useRef } from "react";
+import { logger } from "matrix-js-sdk/src/logger";
+import {
+    IRequestMsisdnTokenResponse,
+    IRequestTokenResponse,
+    MatrixError,
+    ThreepidMedium,
+} from "matrix-js-sdk/src/matrix";
+
+import AddThreepid, { Binding, ThirdPartyIdentifier } from "../../../AddThreepid";
+import { _t, UserFriendlyError } from "../../../languageHandler";
+import AccessibleButton, { ButtonEvent } from "../elements/AccessibleButton";
+import { useMatrixClientContext } from "../../../contexts/MatrixClientContext";
+import Modal from "../../../Modal";
+import ErrorDialog, { extractErrorMessageFromError } from "../dialogs/ErrorDialog";
+import Field from "../elements/Field";
+import { looksValid as emailLooksValid } from "../../../email";
+import CountryDropdown from "../auth/CountryDropdown";
+import { PhoneNumberCountryDefinition } from "../../../phonenumber";
+import InlineSpinner from "../elements/InlineSpinner";
+
+// Whether we're adding 3pids to the user's account on the homeserver or sharing them on an identity server
+type TheepidControlMode = "hs" | "is";
+
+interface ExistingThreepidProps {
+    mode: TheepidControlMode;
+    threepid: ThirdPartyIdentifier;
+    onChange: (threepid: ThirdPartyIdentifier) => void;
+    disabled?: boolean;
+}
+
+const ExistingThreepid: React.FC<ExistingThreepidProps> = ({ mode, threepid, onChange, disabled }) => {
+    const [isConfirming, setIsConfirming] = React.useState(false);
```

Comment:
Any reason for not importing `useState` directly? it is the far more common approach in our codebase